src/server/util/audit.js

```diff
@@ -0,0 +1,57 @@
+import { models } from '../models'
+import { Address4, Address6 } from 'ip-address'
+
+export function getIpVersionAndAddress(ip) {
+  // if it does not contain ':', it's probably an ipv4 address
+  if (!ip.includes(':')) {
+    return ['ipv4', new Address4(ip)]
+  }
+
+  // if it includes a ':' and a '.', is's probably a "IPv4-mapped IPv6 addresses"
+  // see https://en.wikipedia.org/wiki/IPv6#IPv4-mapped_IPv6_addresses for details
+  if (ip.includes('.')) {
+    const ipv4 = ip.split(':').pop()
+    return ['ipv4', new Address4(ipv4)]
+  }
+
+  // otherwise it's probably a proper v6 address
+  return ['ipv6', new Address6(ip)]
+}
+
+async function findOrCreateClientData(req) {
+  const userAgent = req.headers['user-agent'] || ''
+  const [ipVersion, ipAddress] = getIpVersionAndAddress(req.ip)
+
+  const [data] = await models.AuditClientData.findOrCreate({
+    where: {
+      userAgent,
+      ipAddress: ipAddress.bigInteger(),
+      ipVersion,
+    },
+  })
+
+  return data
+}
+
+export async function audit({
+  req, // express request object
```

Comment:
I think passing in req to this audit function is a bit of a layering issue. It pushes the outer edge (express.js), all the way in with the database.

A better model might be to accept the raw data that is needed out of the request (user id, user agent, ip) and create a helper function that can be used to extract all that out of the request. Then the express route handlers would use that function, and pass the data into audit, or any function in between (like the mass edit function), and only the route handlers would need to care about the structure of the req object

```js
audit({ ...getRequestInfo(req), modelId, ... })
```

Reply:
I had already thought of that, just forgot to actually change this :D Anyways, done now.